File: v16_bridge/java/RemoteTV.java
package v16_bridge.java;

abstract class RemoteControl {
    protected Device device;

    public RemoteControl(Device device) {
        this.device = device;
    }

    public abstract void powerOn();

    public abstract void powerOff();

    public abstract void volumeUp();

    public abstract void volumeDown();
}

interface Device {
    void powerOn();

    void powerOff();

    void setVolume(int volume);

    int getVolume();
}

class TV implements Device {
    private int volume = 50;

    @Override
    public void powerOn() {
        System.out.println("TV is on.");
    }

    @Override
    public void powerOff() {
        System.out.println("TV is off.");
    }

    @Override
    public void setVolume(int volume) {
        if (volume > 100) {
            this.volume = 100;
        } else if (volume < 0) {
            this.volume = 0;
        } else {
            this.volume = volume;
        }
        System.out.println("TV volume set to " + this.volume);
    }

    @Override
    public int getVolume() {
        return volume;
    }
}

class DVDPlayer implements Device {
    private int volume = 50;

    @Override
    public void powerOn() {
        System.out.println("DVD player is on.");
    }

    @Override
    public void powerOff() {
        System.out.println("DVD player is off.");
    }

    @Override
    public void setVolume(int volume) {
        if (volume > 100) {
            this.volume = 100;
        } else if (volume < 0) {
            this.volume = 0;
        } else {
            this.volume = volume;
        }
        System.out.println("DVD player volume set to " + this.volume);
    }

    @Override
    public int getVolume() {
        return volume;
    }
}

class BasicRemote extends RemoteControl {
    public BasicRemote(Device device) {
        super(device);
    }

    @Override
    public void powerOn() {
        System.out.print("Basic Remote: power on -> ");
        device.powerOn();
    }

    @Override
    public void powerOff() {
        System.out.print("Basic Remote: power off -> ");
        device.powerOff();
    }

    @Override
    public void volumeUp() {
        device.setVolume(device.getVolume() + 10);
    }

    @Override
    public void volumeDown() {
        device.setVolume(device.getVolume() - 10);
    }
}

class AdvancedRemote extends RemoteControl {
    public AdvancedRemote(Device device) {
        super(device);
    }

    @Override
    public void powerOn() {
        System.out.print("Advanced Remote: power on -> ");
        device.powerOn();
    }

    @Override
    public void powerOff() {
        System.out.print("Advanced Remote: power off -> ");
        device.powerOff();
    }

    @Override
    public void volumeUp() {
        device.setVolume(device.getVolume() + 5);
    }

    @Override
    public void volumeDown() {
        device.setVolume(device.getVolume() - 5);
    }
}

public class RemoteTV {

    public static void main(String[] args) {
        Device tv = new TV();
        Device dvdPlayer = new DVDPlayer();

        RemoteControl basicRemote = new BasicRemote(tv);
        basicRemote.powerOn();
        basicRemote.powerOff();
        basicRemote.volumeUp();
        basicRemote.volumeDown();

        RemoteControl advancedRemote = new AdvancedRemote(dvdPlayer);
        advancedRemote.powerOn();
        advancedRemote.powerOff();
        advancedRemote.volumeUp();
        advancedRemote.volumeDown();

        basicRemote = new BasicRemote(dvdPlayer);
        basicRemote.powerOn();
        basicRemote.powerOff();
        basicRemote.volumeUp();
        basicRemote.volumeDown();
    }
}

File: v16_bridge/python/remote_tv.py
from abc import ABC, abstractmethod

class RemoteControl(ABC):
    def __init__(self, device):
        self.device = device

    @abstractmethod
    def powerOn(self):
        pass

    @abstractmethod
    def powerOff(self):
        pass

    @abstractmethod
    def volumeUp(self):
        pass

    @abstractmethod
    def volumeDown(self):
        pass

class Device(ABC):
    @abstractmethod
    def powerOn(self):
        pass

    @abstractmethod
    def powerOff(self):
        pass

    @abstractmethod
    def setVolume(self, volume):
        pass

    @abstractmethod
    def getVolume(self):
        pass

class TV(Device):
    def __init__(self):
        self.volume = 50

    def powerOn(self):
        print("TV is on.")

    def powerOff(self):
        print("TV is off.")

    def setVolume(self, volume):
        if volume > 100:
            self.volume = 100
        elif volume < 0:
            self.volume = 0
        else:
            self.volume = volume
        print("TV volume set to", self.volume)

    def getVolume(self):
        return self.volume

class DVDPlayer(Device):
    def __init__(self):
        self.volume = 50

    def powerOn(self):
        print("DVD player is on.")

    def powerOff(self):
        print("DVD player is off.")

    def setVolume(self, volume):
        if volume > 100:
            self.volume = 100
        elif volume < 0:
            self.volume = 0
        else:
            self.volume = volume
        print("DVD player volume set to", self.volume)

    def getVolume(self):
        return self.volume

class BasicRemote(RemoteControl):
    def __init__(self, device):
        super().__init__(device)

    def powerOn(self):
        print("Basic Remote: power on ->", end=" ")
        self.device.powerOn()

    def powerOff(self):
        print("Basic Remote: power off ->", end=" ")
        self.device.powerOff()

    def volumeUp(self):
        self.device.setVolume(self.device.getVolume() + 10)

    def volumeDown(self):
        self.device.setVolume(self.device.getVolume() - 10)

class AdvancedRemote(RemoteControl):
    def __init__(self, device):
        super().__init__(device)

    def powerOn(self):
        print("Advanced Remote: power on ->", end=" ")
        self.device.powerOn()

    def powerOff(self):
        print("Advanced Remote: power off ->", end=" ")
        self.device.powerOff()

    def volumeUp(self):
        self.device.setVolume(self.device.getVolume() + 5)

    def volumeDown(self):
        self.device.setVolume(self.device.getVolume() - 5)


if __name__ == "__main__":
    tv = TV()
    dvdPlayer = DVDPlayer()

    basicRemote = BasicRemote(tv)
    basicRemote.powerOn()
    basicRemote.powerOff()
    basicRemote.volumeUp()
    basicRemote.volumeDown()

    advancedRemote = AdvancedRemote(dvdPlayer)
    advancedRemote.powerOn()
    advancedRemote.powerOff()
    advancedRemote.volumeUp()
    advancedRemote.volumeDown()

    basicRemote = BasicRemote(dvdPlayer)
    basicRemote.powerOn()
    basicRemote.powerOff()
    basicRemote.volumeUp()
    basicRemote.volumeDown()


